src/encoder.rs

```diff
@@ -75,6 +75,23 @@ impl<W: Write> Encoder<W> {
         self.info.trns = Some(trns);
     }
 
+    pub fn set_source_gamma(&mut self, source_gamma: f32) {
+        assert!(source_gamma > 0.);
```

Comment:
The idea is to notify the user at the point where there is a mismatch between definition and the actual values. Since gamma cannot have any value that matches the u32 / 100000 specification given by the format, the assertion should actually be much stronger still: it should assert whether a round-trip conversion is possible and fail otherwise, Alternatively, the method should only accept an u32 but this will confuse the user who is expecting some value like 1/1.8 or 1/2.2 to simply work. We'll probably end up with a solution like the ScaledFloat above.

Reply:
Accepting ScaledFloat would be perfect. With the newtype you could have a query of sorts on that type instead of panicking here. Note that the rest of the pipeline already doesn't permit _exact_ roundtrip due to the floating point division by something that isn't a power of two. We had similar issues with time offsets between frames in APNG and decided that the most consistent behaviour would be to approximate values on such conversion. Clamping to the allowed range is in line with that approximation.

```rust
impl ScaledFloat {
    fn in_range(val: f32) -> bool;
    fn exact(val: f32) -> bool;
}
```